### PR TITLE
[core] Improve the workflow finding Redis leader.

### LINKED ISSUE #33744: [Core] Better handling of redis errors
### What happened + What you expected to happen

When we use ray with external redis, sometimes redis returns REDIS_REPLY_ERROR and we are not handle it properly.

### Versions / Dependencies

master

### Reproduction script

#33733

### Issue Severity

None

### PR DESCRIPTION
**Pull request by fishbone**
Signed-off-by: Yi Cheng <74173148+iycheng@users.noreply.github.com>

## Why are these changes needed?
The current way of finding Redis leader sometimes giving error information. The main reason is because the ip address is not resolved.

If in the initialization stage, it connect to the master, it'll use passed in address as the leader which later might make Ray pick follower redis.

This PR fixed the issue and also uses the way redis-cli used to pick the leader. 

The PR makes the checking more strict to give better error message. New protocol as below:

- use boost to resolve the ip address from domain name.
- connect to the first ip address
- if it's cluster mode, 
  - make sure it's healthy; make sure only 1 shard
  - send a dummy write and check the return
    - if return OK, use the ip address directly
    - otherwise, use the one mentioned in the error message
- if not cluster mode, just use the ip address

Refactoring is also done in this PR. Moving connection related information to redis context

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/gcs/redis_context.cc [hunk 1]
```diff
@@ -384,6 +388,63 @@ Status RedisContext::PingPort(const std::string &address, int port) {
...
+  // cluster_size:1
+  RAY_CHECK(reply && !reply->IsNil()) << "Failed to get Redis cluster info";
+  auto cluster_info = reply->ReadAsString();
+
+  std::vector<std::string> parts = absl::StrSplit(cluster_info, "\r\n");
+  bool cluster_mode = false;
+  int cluster_size = 0;
+
+  // Check the cluster status first
+  for (auto &part : parts) {
```
Comment (rkooo567) [inline src/ray/gcs/redis_context.cc:406]:
```suggestion
  for (const auto &part : parts) {
```
nit

File: src/ray/gcs/redis_context.cc [hunk 2]
```diff
@@ -384,6 +388,63 @@ Status RedisContext::PingPort(const std::string &address, int port) {
...
+    if (part.empty() || part[0] == '#') {
```
Comment (rkooo567) [inline src/ray/gcs/redis_context.cc:407]: Is there any way to test this method with real redis? Is it in CI test?

File: src/ray/gcs/redis_context.cc [hunk 3]
```diff
@@ -384,6 +388,63 @@ Status RedisContext::PingPort(const std::string &address, int port) {
...
+      continue;
+    }
+    std::vector<std::string> kv = absl::StrSplit(part, ":");
+    RAY_CHECK(kv.size() == 2);
+    if (kv[0] == "cluster_state") {
+      if (kv[1] == "ok") {
+        cluster_mode = true;
+      } else if (kv[1] == "fail") {
+        RAY_LOG(FATAL)
+            << "The Redis cluster is not healthy. cluster_state shows failed status: "
```
Comment (rkooo567) [inline src/ray/gcs/redis_context.cc:418]: Not sure if it is possible, but can we improve the comment to add an action item in this case?

File: src/ray/gcs/redis_context.cc [hunk 4]
```diff
@@ -384,6 +388,63 @@ Status RedisContext::PingPort(const std::string &address, int port) {
...
+      }
+    }
+    if (kv[0] == "cluster_size") {
+      cluster_size = std::stoi(kv[1]);
+    }
+  }
+
+  if (cluster_mode) {
+    RAY_CHECK(cluster_size == 1)
+        << "Ray currently doesn't support sharded Redis cluster. "
```
Comment (rkooo567) [inline src/ray/gcs/redis_context.cc:429]:
```suggestion
        << "The Redis cluster size is " << cluster_size << ". Ray currently doesn't support sharded Redis cluster. "
```

Comment (rkooo567) [reply src/ray/gcs/redis_context.cc:429]: Maybe I am ignorant in Redis, but "not supporting shared cluster, and provide cluster with 1 shard" sounds contradicting. Does this mean we can have only 1 shard cluster? (master + 1 shard)

File: src/ray/gcs/redis_context.cc [hunk 5]
```diff
@@ -384,6 +388,63 @@ Status RedisContext::PingPort(const std::string &address, int port) {
...
+        << "Please make sure the provided Redis cluster only has one shard";
+  }
+}
+
+std::vector<std::string> ResolveDNS(const std::string &address, int port) {
```
Comment (rkooo567) [inline src/ray/gcs/redis_context.cc:434]: comment to use only the first ip address? Btw, when is the return value has size > 1?

**Review by rkooo567 — APPROVED** [review]
LGTM. Mostly nits

File: src/ray/gcs/redis_context.cc (hunk 4 above)
Comment (iycheng) [reply src/ray/gcs/redis_context.cc:429]: Oh, 1 shard and for each shard you still can have multiple replicas.

File: src/ray/gcs/redis_context.cc (hunk 2 above)
Comment (iycheng) [reply src/ray/gcs/redis_context.cc:407]: We have redis e2e test for this one. Do you mean testing the parsing part only ?

File: src/ray/gcs/redis_context.cc [hunk 6]
```diff
@@ -412,6 +478,39 @@ Status RedisContext::Connect(const std::string &address,
...
+    freeReplyObject(redis_reply);
+    std::vector<std::string> parts = absl::StrSplit(error_msg, " ");
+    RAY_CHECK(parts[0] == "MOVED" && parts.size() == 3)
+        << "Setup Redis cluster failed in the dummy deletion: " << error_msg;
+    std::vector<std::string> ip_port = absl::StrSplit(parts[2], ":");
+    RAY_CHECK(ip_port.size() == 2);
+
+    Disconnect();
+    // Connect to the true leader.
+    return Connect(ip_port[0], std::stoi(ip_port[1]), sharding, password, enable_ssl);
```
[the remark this answers is not in the document]
Comment (iycheng) [reply src/ray/gcs/redis_context.cc:509]: We need this.

**Comment by iycheng** [comment]
@rkooo567 sometimes the user will pass a domain name, for example, MemoryDB it provides a domain name. You'll get several ip address behind that. Some are replicas and some are the leader of the shard.

If the first connection connects to the leader, it's OK and otherwise, we need to switch to the leader.
If it's not the leader, we'll get MOVED error and the right leader will be put in the msg.

Redis can have multiple shards and each shard can have several replicas. Right now, we only support 1 shard the whole redis cluster.

File: src/ray/gcs/redis_context.cc [hunk 7]
```diff
@@ -278,6 +281,7 @@ RedisContext::~RedisContext() {
     redisFreeSSLContext(ssl_context_);
     ssl_context_ = nullptr;
   }
+  redis_async_context_.reset();
```
[the remark this answers is not in the document]
Comment (iycheng) [reply src/ray/gcs/redis_context.cc:284]: it'll be used from other threads, but it's thread safe. ([link](https://sourcegraph.com/github.com/ray-project/ray@95be0b33e898b333a76e84cce8921cec38c097d5/-/blob/src/ray/gcs/redis_async_context.cc?L101))

File: src/ray/gcs/redis_context.cc [hunk 8]
```diff
@@ -412,6 +478,39 @@ Status RedisContext::Connect(const std::string &address,
...
+    argv.push_back(arg.data());
+    argc.push_back(arg.size());
+  }
+
+  auto redis_reply = reinterpret_cast<redisReply *>(
+      ::redisCommandArgv(context_, cmds.size(), argv.data(), argc.data()));
+
+  if (redis_reply->type == REDIS_REPLY_ERROR) {
+    // This should be a MOVED error
+    // MOVED 14946 10.xx.xx.xx:7001
```
[the remark this answers is not in the document]
Comment (scv119) [reply src/ray/gcs/redis_context.cc:498]: wondering if this message can change in the future, and is there a way for us to detect that?

File: src/ray/gcs/redis_context.cc (hunk 5 above)
Comment (iycheng) [reply src/ray/gcs/redis_context.cc:434]: Trying to make this method generic. I'll put comments in the caller side.

File: src/ray/gcs/redis_context.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (rkooo567) [reply src/ray/gcs/redis_context.cc:284]: So for other code path, they all handle nullptr redis_async_context_ right?

**Comment by rkooo567** [comment]
> @rkooo567 sometimes the user will pass a domain name, for example, MemoryDB it provides a domain name. You'll get several ip address behind that. Some are replicas and some are the leader of the shard.

> If the first connection connects to the leader, it's OK and otherwise, we need to switch to the leader.
If it's not the leader, we'll get MOVED error and the right leader will be put in the msg.

Make sense. It'd be great to add this comment to the code!

File: src/ray/gcs/redis_context.cc [hunk 9]
```diff
@@ -384,15 +388,92 @@ Status RedisContext::PingPort(const std::string &address, int port) {
...
+  }
+  return ip_addresses;
+}
+
 Status RedisContext::Connect(const std::string &address,
                              int port,
                              bool sharding,
                              const std::string &password,
                              bool enable_ssl) {
+  // Connect to the leader of the Redis cluster:
```
Comment (iycheng) [inline src/ray/gcs/redis_context.cc:453]: @rkooo567  comments has been added here.

**Pull request merged**